### PR TITLE
Ocd 969

### PR DESCRIPTION
**Pull request by tmy1313**

### CODE REVIEW

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java [hunk 1]
```diff
@@ -97,7 +97,10 @@
     @Autowired private MessageSource messageSource;
 
     @ApiOperation(value = "Get auditable data for certification bodies.",
-            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results.")
+            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results. "
+                    + "Security Restrictions: ROLE_ADMIN AND ROLE_ONC_ADMIN may set the 'showDeleted' flag to true "
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: I thought we got rid of the `showDeleted` flags?

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java [hunk 2]
```diff
@@ -565,7 +580,9 @@
     }
 
     @ApiOperation(value = "Get auditable data about all CHPL user accounts",
-            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results.")
+            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results.  "
+                    + "Security Restrictions: ROLE_ADMIN, ROLE_ONC_ADMIN, ROLE_ONC_STAFF (of ROLE_ONC_STAFF), ROLE_CMS_STAFF "
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: ROLE_ONC_ADMIN never existed, and ROLE_ONC_STAFF was rolled into ROLE_ONC

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationBodyController.java [hunk 3]
```diff
@@ -53,14 +52,14 @@
 
     @ApiOperation(value = "List all certification bodies (ACBs).",
             notes = "Setting the 'editable' parameter to true will return all ACBs that the logged in user has "
-                    + "edit permissions on. The logged in user must have ROLE_ADMIN or ROLE_ONC to see retired ACBs."
-                    + "The default behavior of this service is to list all of the ACBs in the system that are not retired.")
+                    + "edit permissions on. Security Restrictions:  Everyone can see all active.  ROLE_ADMIN and "
+                    + "ROLE_ONC_ADMIN can see deleted.")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationBodyController.java]: I don't think anyone can see deleted any more

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java (hunk 1 above)
Comment (kekey1) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: We did and I got rid of the showDeleted documentation at that time.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationBodyController.java (hunk 3 above)
Comment (kekey1) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationBodyController.java]: I think the previous documentation was correct except maybe that it did not specifically call out anonymous users.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationIdController.java [hunk 4]
```diff
@@ -57,8 +57,9 @@
     //
     // Retrieves all CMS Certification IDs and their date of creation.
     // **********************************************************************************************************
-    @PreAuthorize("hasAnyRole('ROLE_ADMIN', 'ROLE_ONC', 'ROLE_CMS_STAFF')")
-    @ApiOperation(value = "Retrieves a list of all CMS EHR Certification IDs along with the date they were created.")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN', 'ROLE_ONC', 'ROLE_CMS_STAFF', 'ROLE_ONC_STAFF')")
```
Comment (kekey1) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationIdController.java]: Didn't we get rid of ROLE_ONC_STAFF? I think the PreAuthorize was correct.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java (hunk 1 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: I was simply cutting and pasting the security restrictions from the spreadhseet.  The code is correct and the  Swagger annotations should be correct now.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java (hunk 2 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: ROLE_ONC_ADMIN was in the spreadsheet.  Went though and replaced all with ROLE_ONC.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationBodyController.java (hunk 3 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationBodyController.java]: I was simply cutting and pasting the security restrictions from the spreadsheet.  The code is correct and the  Swagger annotations should be correct now.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationIdController.java (hunk 4 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationIdController.java]: Not sure what I did here!  Should be fixed, though.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java [hunk 5]
```diff
@@ -664,8 +680,8 @@
     }
 
     @ApiOperation(value = "Track the actions of all users in the system",
-            notes = "The authenticated user calling this method must have ROLE_ADMIN or ROLE_ONC. "
-                    + "Users must specify 'start' and 'end' parameters to restrict the date range of the results.")
+            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results."
+                    + "Security Restrictions: ROLE_ADMIN, ROLE_ONC, or ROLE_ONC_STAFF.")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: ROLE_ONC_STAFF is gone

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java [hunk 6]
```diff
@@ -678,8 +694,8 @@
     }
 
     @ApiOperation(value = "Track the actions of a specific user in the system",
-            notes = "The authenticated user calling this method must have ROLE_ADMIN or ROLE_ONC. "
-                    + "A start and end date may optionally be provided to limit activity results.")
+            notes = "A start and end date may optionally be provided to limit activity results."
+                    + "Security Restrictions: ROLE_ADMIN, ROLE_ONC, or ROLE_ONC_STAFF.")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: Also

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java [hunk 7]
```diff
@@ -60,9 +57,12 @@
         return results;
     }
 
-    @ApiOperation(value = "Get a specific announcement.")
+    @ApiOperation(value = "Get a specific announcement.",
+            notes = "Security Restrictions: ROLE_ADMIN, ROLE_ONC can retrieve future scheduled "
+                    + "announcements and private announcements.  ROLE_ONC_STAFF, ROLE_ACB, ROLE_ATL, and ROLE_CMS_STAFF "
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java]: ROLE_ONC_STAFF

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java [hunk 8]
```diff
@@ -71,17 +71,17 @@
     }
 
     @ApiOperation(value = "Create a new announcement.",
-            notes = "Only CHPL users with ROLE_ADMIN and ROLE_ONC are able to create announcements.")
+            notes = "Security Restrictions: ROLE_ADMIN, ROLE_ONC, or ROLE_ONC_STAFF")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java]: ROLE_ONC STAFF

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ApiKeyController.java [hunk 9]
```diff
@@ -103,7 +104,7 @@ private String delete(final String key, final String userApiKey, final String us
     }
 
     @ApiOperation(value = "List all API keys that have been created.",
-            notes = "This service is only available to CHPL users with ROLE_ADMIN or ROLE_ONC.")
+            notes = "Security Restrictions: ROLE_ADMIN, ROLE_ONC, or ROLE_ONC_STAFF")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ApiKeyController.java]: Should probably Ctrl-F this value. I'm going to stop commenting

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertifiedProductController.java [hunk 10]
```diff
@@ -555,11 +555,11 @@ public void streamSEDDetailsDocumentContents(final HttpServletResponse response)
...
     @ApiOperation(value = "Update an existing certified product.",
-            notes = "Updates the certified product after first validating the request. The logged in"
-                    + " user must have ROLE_ADMIN, ROLE_ONC, or ROLE_ACB and have administrative "
-                    + " authority on the ACB that certified the product. If a different ACB is passed in"
-                    + " as part of the request, an ownership change will take place and the logged in "
-                    + " user must have ROLE_ADMIN or ROLE_ONC.")
+            notes = "Updates the certified product after first validating the request. If a different "
+                    + "ACB is passed in as part of the request, an ownership change will take place and "
+                    + "the logged in user must have ROLE_ADMIN or ROLE_ONC.  Security Restrictions: ROLE_ADMIN, "
+                    + "ROLE_ONC, or ROLE_ACB and have administrative authority on the ACB that certified "
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertifiedProductController.java]: Can a listing be transferred from one ACB to another by a user that is ROLE_ACB? If so, do they need ACB admin on both ACBs? If so, that's not explicit here. It looks like they only need it on the originating ACB

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SchedulerController.java [hunk 11]
```diff
@@ -118,7 +124,8 @@ public void deleteTrigger(@PathVariable("triggerGroup") final String triggerGrou
      * @return the updated job
      * @throws SchedulerException if scheduler has issues
      */
-    @ApiOperation(value = "Update a given job")
+    @ApiOperation(value = "Update a given job",
+            notes = "Security Restrictions: ROLE_ADMIN or ROLE_ACB and have administrative authority on the specified ACB.")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SchedulerController.java]: The only job that can be updated is the one that's only available to ROLE_ADMIN and ROLE_ONC, right? I don't think ROLE_ACB has any Jobs they can update

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SchedulerController.java [hunk 12]
```diff
@@ -40,7 +40,8 @@
      * @throws SchedulerException if exception is thrown
      * @throws ValidationException if job values aren't correct
      */
-    @ApiOperation(value = "Create a new trigger and return it")
+    @ApiOperation(value = "Create a new trigger and return it",
+            notes = "Security Restrictions: ROLE_ADMIN or ROLE_ACB and have administrative authority on the specified ACB.")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SchedulerController.java]: This one should have ROLE_ONC too, right? Also think there's something odd about the ACB, because not all triggers specify an ACB. Similar comments to the other endpoints related to triggers & ACB

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/TestingLabController.java [hunk 13]
```diff
@@ -57,9 +56,8 @@
 
     @ApiOperation(value = "List all testing labs (ATLs).",
             notes = "Setting the 'editable' parameter to true will return all ATLs that the logged in user has edit "
-                    + "permissions on.  Setting 'showDeleted' to true will include even those ATLs that have been "
-                    + "deleted. The logged in user must have ROLE_ADMIN to see deleted ATLs. The default behavior of "
-                    + "this service is to list all of the ATLs in the system that are not deleted.")
+                    + "permissions on.  Security Restrictions: ROLE_ADMIN or ROLE_ONC can see all ATLs.  ROLE_ATL "
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/TestingLabController.java]: This phrasing implies that anon cannot call this method, but this is called to populate the /overview page, so I know it works there

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/TestingLabController.java [hunk 14]
```diff
@@ -80,8 +78,8 @@
     }
 
     @ApiOperation(value = "Get details about a specific testing lab (ATL).",
-            notes = "The logged in user must have ROLE_ADMIN or have either read or"
-                    + "administrative authority on the testing lab with the ID specified.")
+            notes = "Security Restrictions: ROLE_ADMIN, ROLE_ONC or have either read or administrative authority "
+                    + "on the testing lab with the ID specified.")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/TestingLabController.java]: Is this really true? Can the public call `/atls` and get all the data for all ATLs, but can't call `/atl/1`?

File: chpl/chpl-auth/src/main/java/gov/healthit/chpl/auth/entity/UserEntity.java [hunk 15]
```diff
@@ -73,6 +75,11 @@
     @ManyToOne(optional = false, fetch = FetchType.EAGER)
     @JoinColumn(name = "contact_id", unique = true, nullable = false)
     private UserContactEntity contact;
+    
+    @ManyToMany
+    @JoinTable(name = "user_certification_body_map",
+            joinColumns = { @JoinColumn(name = "user_id") })
```
Comment (andlar) [inline chpl/chpl-auth/src/main/java/gov/healthit/chpl/auth/entity/UserEntity.java]: I don't think this file should be part of this PR

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/entity/UserCertificationBodyMapEntity.java [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.OneToOne;
+import javax.persistence.Table;
+
+import gov.healthit.chpl.auth.entity.UserEntity;
+
+@Entity
+@Table(name = "user_certification_body_map")
+public class UserCertificationBodyMapEntity {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/entity/UserCertificationBodyMapEntity.java]: Also not part of this PR

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/impl/CertifiedProductManagerImpl.java [hunk 17]
```diff
@@ -1163,31 +1155,34 @@ public CertifiedProductDTO update(final Long acbId, final ListingUpdateRequest u
...
                             .equals(CertificationStatusType.TerminatedByOnc.toString())) {
-                        newDevStatusDto = devStatusDao.getByName(DeveloperStatusType.UnderCertificationBanByOnc.toString());
+                        newDevStatusDto = devStatusDao
+                                .getByName(DeveloperStatusType.UnderCertificationBanByOnc.toString());
                     }
                 } else if (!Util.isUserRoleAdmin() && !Util.isUserRoleOnc()) {
                     LOGGER.error("User " + Util.getUsername()
-                    + " does not have ROLE_ADMIN and cannot change the status of developer for certified product with id "
-                    + listingId);
+                            + " does not have ROLE_ADMIN and cannot change the status of developer for certified product with id "
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/impl/CertifiedProductManagerImpl.java]:
```suggestion
                            + " does not have ROLE_ADMIN or ROLE_ONC and cannot change the status of developer for certified product with id "
```

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java (hunk 5 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java (hunk 6 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java (hunk 7 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java (hunk 8 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ApiKeyController.java (hunk 9 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ApiKeyController.java]: Addressed - plus all others...

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SchedulerController.java (hunk 11 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SchedulerController.java]: Addressed

File: chpl/chpl-auth/src/main/java/gov/healthit/chpl/auth/entity/UserEntity.java (hunk 15 above)
Comment (tmy1313) [reply chpl/chpl-auth/src/main/java/gov/healthit/chpl/auth/entity/UserEntity.java]: Addressed

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/entity/UserCertificationBodyMapEntity.java (hunk 16 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/entity/UserCertificationBodyMapEntity.java]: Addressed

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/impl/CertifiedProductManagerImpl.java (hunk 17 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/impl/CertifiedProductManagerImpl.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/TestingLabController.java (hunk 13 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/TestingLabController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/TestingLabController.java (hunk 14 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/TestingLabController.java]: There is no security in the manager.  Documentation updated.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SchedulerController.java (hunk 12 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SchedulerController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertifiedProductController.java (hunk 10 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertifiedProductController.java]: Updated the description to match the manager's security.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/UserManagementController.java [hunk 18]
```diff
@@ -380,7 +383,7 @@ private User update(final User userInfo)
 
     @ApiOperation(value = "Delete a user.",
             notes = "Deletes a user account and all associated authorities on ACBs and ATLs. "
-                    + "The logged in user must have ROLE_ADMIN.")
+                    + "Security Restrictions: ROLE_ADMIN or ROLE_ONC_ADMIN")
```
Comment (andlar) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/UserManagementController.java]: I don't think ROLE_ONC_ADMIN exists

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java [hunk 19]
```diff
@@ -97,7 +97,9 @@
     @Autowired private MessageSource messageSource;
 
     @ApiOperation(value = "Get auditable data for certification bodies.",
-            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results.")
+            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results. "
+                    + "Security Restrictions: ROLE_ADMIN AND ROLE_ONC may see activity for all certification bodies.  "
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java [hunk 20]
```diff
@@ -271,7 +280,8 @@
     }
 
     @ApiOperation(value = "Get auditable data for all API keys",
-            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results.")
+            notes = "Users must specify 'start' and 'end' parameters to restrict the date range of the results.  "
+                    + "Security Restrictions: Only ROLE_ADMIN & ROLE_ONC")
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ActivityController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java [hunk 21]
```diff
@@ -71,17 +71,17 @@
     }
 
     @ApiOperation(value = "Create a new announcement.",
-            notes = "Only CHPL users with ROLE_ADMIN and ROLE_ONC are able to create announcements.")
+            notes = "Security Restrictions: ROLE_ADMIN or ROLE_ONC")
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/AnnouncementController.java]: Addressed

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationBodyController.java [hunk 22]
```diff
@@ -76,8 +74,7 @@
     }
 
     @ApiOperation(value = "Get details about a specific certification body (ACB).",
-            notes = "The logged in user must either have ROLE_ADMIN or have ROLE_ACB "
-                    + " for the ACB with the provided ID.")
+            notes = "Security Restriction:  ROLE_ADMIN, ROLE_ONC, or ROLE_ACB for the ACB with the provided ID")
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertificationBodyController.java]: There is no security in the underlying manager - addressed.

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/UserManagementController.java (hunk 18 above)
Comment (tmy1313) [reply chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/UserManagementController.java]: Addressed
